codex/sales/slotqueue.nim

```diff
@@ -332,6 +363,23 @@ proc dispatch(self: SlotQueue,
       # throw because it is caller-defined
       warn "Unknown error processing slot in worker", error = e.msg
 
+proc clearSeenFlags*(self: SlotQueue) =
+  # Enumerate all items in the queue, overwriting each item with `seen = true`.
+  # To avoid issues with new queue items being pushed to the queue while all
+  # items are being iterated (eg if a new storage request comes in and pushes
+  # new slots to the queue), this routine must remain synchronous.
+
+  if self.queue.empty:
+    return
+
+  for item in self.queue.mitems:
+    item.seen = false # does not maintain the heap invariant
+
+  # force heap reshuffling to maintain the heap invariant
+  doAssert self.queue.update(self.queue[0]), "slot queue failed to reshuffle"
```

Comment:
`update()` only returns false when the argument is not in the queue. Since we're supplying `self.queue[0]` as the arguments, this will never happen. And if it does happen, then it's a programmer error and the node should crash, I think.

Reply:
Yes, exactly. In addition, we know that self.queue[0] exists because there is an early return
when `if self.queue.empty` is true. If the queue is empty when we don't expect
it to be, there is no real way to recover.